src/transports/tcp.rs

```diff
@@ -410,32 +419,58 @@ impl<Id: PeerId> Transport<Id> for TcpTransport<Id> {
     }
 
     fn send(endpoint: &mut Self::Endpoint, data: &[u8]) -> PeerNetResult<()> {
+        let start_time = Instant::now();
         let msg_size: u32 = data.len().try_into().map_err(|_| {
             TcpError::ConnectionError
                 .wrap()
                 .error("send len too long", Some(format!("{:?}", data.len())))
         })?;
-        if msg_size > 1048576000 {
+        if msg_size > endpoint.config.max_message_size as u32 {
             return Err(TcpError::ConnectionError
                 .wrap()
-                .error("send len too long", Some(format!("{:?}", data.len()))))?;
+                .error("send len too long", Some(format!("{:?}", data.len()))));
+        }
+        // write msg len to stream
+        match endpoint.stream.write(&msg_size.to_be_bytes()) {
+            Ok(0) => {
+                endpoint.shutdown();
+                return Err(TcpError::ConnectionError
+                    .wrap()
+                    .error("send len write = 0", None));
+            }
+            Ok(_count) => {}
+            Err(err) => {
+                return Err(TcpError::ConnectionError
+                    .wrap()
+                    .new("send len write", err, None))
+            }
+        }
+
+        let mut write_count = 0;
+        while write_count < data.len() {
```

Comment:
let's encapsulate all this logic into a `read_exact_timeout(stream: TcpStream, timeout: MassaTIme, read_size: usize) -> Result<Vec<u8>>` method like it was done in Bootstrap

Reply:
done